### ISSUE TITLE
Differing class probabilities between imported and original Keras model

### ISSUE THREAD

**Issue by perceptoid** [deeplearning4j/deeplearning4j#2417]
Simple MNIST testset reproduces issue.

Class probabilities from Keras evaluation.
[7.58e-11 2.77e-09 2.45e-08 5.66e-08 8.83e-11 7.64e-11 2.53e-15 9.99e-01 1.76e-10   5.15e-08]]

Class probabilities from dl4j eval on same image:
[0.06, 0.06, 0.14, 0.18, 0.07, 0.05, 0.04, 0.19, 0.16, 0.05]

Keras backend is Theano.

Warnings when importing model in dl4j:

[main] WARN org.deeplearning4j.nn.modelimport.keras.Model - Changed layer-defined dropout 0.0 to 0.25 because of previous Dropout=0.25 layer
[main] WARN org.deeplearning4j.nn.modelimport.keras.Model - Changed layer-defined dropout 0.0 to 0.5 because of previous Dropout=0.5 layer
[main] WARN org.deeplearning4j.nn.modelimport.keras.LayerConfiguration - DL4J adds reshaping layers during model compilation: https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/MultiLayerConfiguration.java#L429

Attached zip with python code for training and evaluation. Java code for evaluation and jsn/hd5 with model.

[keras_dl4j.zip](https://github.com/deeplearning4j/deeplearning4j/files/624822/keras_dl4j.zip)

**Comment by turambar** [comment]
That is almost certainly because we are improperly interpreting the dropout probabilities. In _Keras_ the dropout probability is the probability of dropping out the units (as shown [here in the TF backend](https://github.com/fchollet/keras/blob/master/keras/backend/tensorflow_backend.py#L1528)...Theano backend same). In DL4J, the dropout parameter represents the probability of _retaining_ the units. The current [model import code](https://github.com/deeplearning4j/deeplearning4j/blob/8ab54a28eb8b4d4683692e34cadf7b4513cf5af7/deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/ModelConfiguration.java#L119) is not accounting for this. Will make a quick bugfix ASAP.

**Comment by turambar** [comment]
Candidate fix pushed to `model_loading` branch ([here](https://github.com/deeplearning4j/deeplearning4j/blob/model_loading/deeplearning4j-modelimport/src/main/java/org/deeplearning4j/nn/modelimport/keras/LayerConfiguration.java)). Won't close this until fully tested, but I encourage you guys to try out the `model_loading` branch if you're able to build from source.

**Comment by turambar** [comment]
Well, this definitely turned up some bugs related to our handling of dropout, which I've fixed. However, those were not the causes of the above behavior (which persists). Since different levels of dropout don't change the output, that suggests that it's most likely a bigger problem with the weights. Will need another day or so to fix.

**Comment by kuluofenghun** [comment]
we are also trying to load a saved keras model, and the model backend is tensorflow. My code is as follow:
import org.deeplearning4j.nn.modelimport.keras.Model;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;


import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class modelimport {
	private static final Logger log = LoggerFactory.getLogger(modelimport.class);

	public static void main(String[] args) throws Exception {
		String jsnPath = "C:\\Users\\liuying\\Desktop\\keras model\\intent_dnn_model_v1.jsn";
		String hdfPath = "C:\\Users\\liuying\\Desktop\\keras model\\intent_dnn_weights_v1.h5";
		MultiLayerNetwork model = Model.importSequentialModel(jsnPath, hdfPath);
		System.out.println("loaded");
	}
}

I build my project in eclipse, java 8. But it is failed,the error information is
![qq 20161206141219](https://cloud.githubusercontent.com/assets/23549708/20915268/c17bc0a8-bbbf-11e6-9331-c0365debb3db.png)
Could you share me some experience and help me solve this question? Thank you !

**Comment by agibsonccc** [comment]
@kuluofenghun  I've already commented that your issue is not a problem with our loader. You are trying to load a file  that doesn't exist.

**Comment by kuluofenghun** [comment]
I am so sorry to trouble you. 
I have correct the file path in last issue. But I still meet some other questions. such as 


I doubt whether my pom.xml file lacks some dependency information. But I can't find this dependency information.
It is really important for my work, I hope you can help me. Thank you so much!  



At 2016-12-07 09:07:06, "Adam Gibson" <notifications@github.com> wrote:


@kuluofenghun I've already commented that your issue is not a problem with our loader. You are trying to load a file that doesn't exist.

**Comment by turambar** [comment]
@perceptoid this is almost certainly fixed in the major update to the [model_loading](https://github.com/deeplearning4j/deeplearning4j/tree/model_loading) branch. However, the fix requires rotating convolutional filters 180 degrees when importing a model built with the Theano backend (as yours was), and there's currently a bug in ND4J that breaks this operation when applied to tensors with shape N x ... x 1 (i.e., images with only one channel).

**Comment by turambar** [comment]
@perceptoid @kuluofenghun @eraly I can confirm that this is fixed in the [kerasModelImport](https://github.com/deeplearning4j/deeplearning4j/tree/kerasModelImport), which will be merged into `master` shortly and available in the next release.
